API_GUIDE.md

```diff
@@ -143,11 +154,23 @@ PyTorch/XLA behaves semantically like regular PyTorch and XLA tensors, implement
     ```
 
 
-3.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch/XLA automatically handles them, but they are seen as different execution graphs and require recompilations.
+1.  Loops with a different number of iterations between steps are subject to similar observations as tensor shapes. PyTorch/XLA automatically handles them, but they are seen as different execution graphs and require recompilations. Therefore, to get the best performance, same computations should run on all XLA devices in all hosts.
```

Comment:
This should be `3.` I think.

Also, there are two different issues.
One is the number of loops which differ across different step.
I would replace the `loops` mentions with one more general along the lines "in order to avoid recompilations, not only shapes must be constant, but also computations" ... and then make the "loop" case be a subcase of that.
That is one issue.

Then you are mixing it with another issue, which is same computation across devices.
There, in training, if the different computation across devices ends up either triggering CRS (Cross Replica Sum) ops with different shapes, or in different sequences, the training would hang.